# Album aggs and Song dataframe Updates based on Billboard Hot 100 Updates


## Based on newly released data (January 1st, 2023) and expanding whole career (vs. August 2019)

## Contents

## 01. Import libraries and datasets
## 02. Wrangling: update all top song categories
## 03. Updating top song counts for album aggs and album scores
## 04. Data export

# 01. Import libraries and data

In [1]:
# Import libraries

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import scipy
import os

In [2]:
# Create path

path = r"C:\Users\leobs\OneDrive\Desktop\LF_TS_Analysis"

In [3]:
# Import song stream for merging song info

df = pd.read_csv(os.path.join(path,'02 Manipulated Data', 'song_stream_alb_final.csv'), index_col = 0)

In [12]:
# Import billboard for comparing value figures


df_bill = pd.read_csv(os.path.join(path,'02 Manipulated Data', 'df_hot_100_ts_only_dec_23.csv'), index_col = 0)

In [49]:
# Import album scores final to update total top songs

df_alb_scores = pd.read_csv(os.path.join(path,'02 Manipulated Data', 'final_album_scores_2.csv'), index_col = 0)

In [50]:
# Import album aggs to update total top songs

df_alb_aggs = pd.read_csv(os.path.join(path,'02 Manipulated Data', 'alb_aggs_final.csv'), index_col = 0)

# 02. Wrangling: update all top song categories within songs dataframe

## Review df and update top songs

In [25]:
pd.set_option('display.max_columns', None)

In [26]:
df.head()

song_title  \
0         Welcome To New York (Taylor's Version)   
1                 Blank Space (Taylor's Version)   
2                       Style (Taylor's Version)   
3            Out Of The Woods (Taylor's Version)   
4  All You Had To Do Was Stay (Taylor's Version)   

                       album_title song_release_date  track      acou  dance  \
0  1989 (Taylor's Version)[Deluxe]        10/27/2023      1  0.009420  0.757   
1  1989 (Taylor's Version)[Deluxe]        10/27/2023      2  0.088500  0.733   
2  1989 (Taylor's Version)[Deluxe]        10/27/2023      3  0.000421  0.511   
3  1989 (Taylor's Version)[Deluxe]        10/27/2023      4  0.000537  0.545   
4  1989 (Taylor's Version)[Deluxe]        10/27/2023      5  0.000656  0.588   

    ener    instru    live   loud  speech     temp    val  pop     dur  \
0  0.610  0.000037  0.3670 -4.840  0.0327  116.998  0.685   80  212600   
1  0.733  0.000000  0.1680 -5.376  0.0670   96.057  0.701   80  231833   
2  0.822  0.019700  0.0899 -4.785  0.0397   94.868  0.305   81  231000   
3  0.885  0.000056  0.3850 -5.968  0.0447   92.021  0.206   80  235800   
4  0.721  0.000000  0.1310 -5.579  0.0317   96.997  0.520   79  193289   

  album_id   song_id  days  tot_streams last_stream_update  streams_n  \
0   1989TV  1989TV01    13     35521714          11/9/2023    2732439   
1   1989TV  1989TV02    13     40958068          11/9/2023    3150620   
2   1989TV  1989TV03    13     47687102          11/9/2023    3668238   
3   1989TV  1989TV04    13     37545128          11/9/2023    2888086   
4   1989TV  1989TV05    13     31008534          11/9/2023    2385271   

   top_song  top_song_flag  top_s_cnt                tv  alb_num  \
0  top song           True         14  Taylor's Version       14   
1  top song           True         14  Taylor's Version       14   
2  top song           True         14  Taylor's Version       14   
3  top song           True         14  Taylor's Version       14   
4  top song           True         14  Taylor's Version       14   

  album_announce_mo  alb_ann_day  alb_ann_yr album_announce_date  \
0            August            9        2023            8/9/2023   
1            August            9        2023            8/9/2023   
2            August            9        2023            8/9/2023   
3            August            9        2023            8/9/2023   
4            August            9        2023            8/9/2023   

  album_release_mo  alb_r_day  alb_r_yr album_release_date  w1_sales  \
0          October         27      2023         10/27/2023   1359000   
1          October         27      2023         10/27/2023   1359000   
2          October         27      2023         10/27/2023   1359000   
3          October         27      2023         10/27/2023   1359000   
4          October         27      2023         10/27/2023   1359000   

   top_alb_count  top_alb_normal  alb_track_cnt  tvflag  \
0              1        0.076923             22    True   
1              1        0.076923             22    True   
2              1        0.076923             22    True   
3              1        0.076923             22    True   
4              1        0.076923             22    True   

             sales_group  tot_streams_group              alb_genre  
0  High first week sales  Low total streams  Synth/Dream/Dance pop  
1  High first week sales  Low total streams  Synth/Dream/Dance pop  
2  High first week sales  Low total streams  Synth/Dream/Dance pop  
3  High first week sales  Low total streams  Synth/Dream/Dance pop  
4  High first week sales  Low total streams  Synth/Dream/Dance pop

In [5]:
df.shape

(285, 42)

In [7]:
df.tail()

song_title album_title song_release_date  \
280                              Our Song       Debut        10/24/2006   
281         I'm Only Me When I'm With You       Debut        10/24/2006   
282                             Invisible       Debut        10/24/2006   
283                A Perfectly Good Heart       Debut        10/24/2006   
284  Teardrops on My Guitar - Pop Version       Debut        10/24/2006   

     track     acou  dance   ener    instru    live   loud  ...  alb_r_yr  \
280     11  0.11100  0.668  0.672  0.000000  0.3290 -4.931  ...      2006   
281     12  0.00452  0.563  0.934  0.000807  0.1030 -3.629  ...      2006   
282     13  0.63700  0.612  0.394  0.000000  0.1470 -5.723  ...      2006   
283     14  0.00349  0.483  0.751  0.000000  0.1280 -5.726  ...      2006   
284     15  0.04020  0.459  0.753  0.000000  0.0863 -3.827  ...      2006   

     album_release_date  w1_sales  top_alb_count  top_alb_normal  \
280          10/24/2006     39000            284        0.045622   
281          10/24/2006     39000            284        0.045622   
282          10/24/2006     39000            284        0.045622   
283          10/24/2006     39000            284        0.045622   
284          10/24/2006     39000            284        0.045622   

    alb_track_cnt tvflag           sales_group   tot_streams_group  \
280            15  False  Low first week sales  High total streams   
281            15  False  Low first week sales   Low total streams   
282            15  False  Low first week sales   Low total streams   
283            15  False  Low first week sales   Low total streams   
284            15  False  Low first week sales   Low total streams   

       alb_genre  
280  Country pop  
281  Country pop  
282  Country pop  
283  Country pop  
284  Country pop  

[5 rows x 42 columns]

In [8]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 285 entries, 0 to 284
Data columns (total 42 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   song_title           285 non-null    object 
 1   album_title          285 non-null    object 
 2   song_release_date    285 non-null    object 
 3   track                285 non-null    int64  
 4   acou                 285 non-null    float64
 5   dance                285 non-null    float64
 6   ener                 285 non-null    float64
 7   instru               285 non-null    float64
 8   live                 285 non-null    float64
 9   loud                 285 non-null    float64
 10  speech               285 non-null    float64
 11  temp                 285 non-null    float64
 12  val                  285 non-null    float64
 13  pop                  285 non-null    int64  
 14  dur                  285 non-null    int64  
 15  album_id             285 non-null    object 


In [9]:
df.columns

Index(['song_title', 'album_title', 'song_release_date', 'track', 'acou',
       'dance', 'ener', 'instru', 'live', 'loud', 'speech', 'temp', 'val',
       'pop', 'dur', 'album_id', 'song_id', 'days', 'tot_streams',
       'last_stream_update', 'streams_n', 'top_song', 'top_song_flag',
       'top_s_cnt', 'tv', 'alb_num', 'album_announce_mo', 'alb_ann_day',
       'alb_ann_yr', 'album_announce_date', 'album_release_mo', 'alb_r_day',
       'alb_r_yr', 'album_release_date', 'w1_sales', 'top_alb_count',
       'top_alb_normal', 'alb_track_cnt', 'tvflag', 'sales_group',
       'tot_streams_group', 'alb_genre'],
      dtype='object')

### Columns to update:

for songs: 'top_song','top_song_flag'
for album top song count: 'top_s_cnt'

In [10]:
# Value counts for top songs

df['top_song'].value_counts(dropna=False)

top_song
top song    192
not top      93
Name: count, dtype: int64

In [14]:
df_bill.head()

chart_week  chart_position song_title        artist  prior_wk_position  \
884  7/27/2013              65         22  Taylor Swift                 55   
885  7/20/2013              55         22  Taylor Swift                 49   
886  7/13/2013              49         22  Taylor Swift                 41   
887   7/6/2013              41         22  Taylor Swift                 32   
889  6/29/2013              32         22  Taylor Swift                 26   

     peak_pos  wks_on_chart           album_title song_release_date  track  \
884        20            20  Red (Deluxe Edition)        10/22/2012      6   
885        20            19  Red (Deluxe Edition)        10/22/2012      6   
886        20            18  Red (Deluxe Edition)        10/22/2012      6   
887        20            17  Red (Deluxe Edition)        10/22/2012      6   
889        20            16  Red (Deluxe Edition)        10/22/2012      6   

     ...  album_release_date  w1_sales  top_alb_count  top_alb_normal  \
884  ...          10/22/2012   1208000            185        0.045849   
885  ...          10/22/2012   1208000            185        0.045849   
886  ...          10/22/2012   1208000            185        0.045849   
887  ...          10/22/2012   1208000            185        0.045849   
889  ...          10/22/2012   1208000            185        0.045849   

     alb_track_cnt  tvflag            sales_group   tot_streams_group  \
884             22   False  High first week sales  High total streams   
885             22   False  High first week sales  High total streams   
886             22   False  High first week sales  High total streams   
887             22   False  High first week sales  High total streams   
889             22   False  High first week sales  High total streams   

     alb_genre  _merge  
884   Pop rock    both  
885   Pop rock    both  
886   Pop rock    both  
887   Pop rock    both  
889   Pop rock    both  

[5 rows x 49 columns]

In [15]:
top_bill_songs = df_bill['song_title'].nunique()

In [17]:
top_bill_songs

213

After update, should have 213 top songs in the songs df

In [18]:
# Change top song flags for list of song ids that are in fact top songs
# Create list to update

song_id_updates = ["EVER04","REDTV30","SPEAK18","198908","1989TV22","FEAR13","SPEAKTV20","EVER09","SPEAKTV17","REP02","EVER15","RED14","REDTV14","FOLK04","REDTV26","MIDN23","LOVE16","EVER06","REDTV23","REDTV28","MIDN04","LOVE12","REDTV10","FEARTV21"]


In [29]:
# Update the 'top_song' flag for the specified song ID

df.loc[df['song_id'].isin(song_id_updates), 'top_song'] = "top song"

In [30]:
# Check value counts

df['top_song'].value_counts(dropna=False)

top_song
top song    216
not top      69
Name: count, dtype: int64

In [31]:
top = df.loc[df['top_song'] == "top song"]

In [32]:
top

song_title  \
0           Welcome To New York (Taylor's Version)   
1                   Blank Space (Taylor's Version)   
2                         Style (Taylor's Version)   
3              Out Of The Woods (Taylor's Version)   
4    All You Had To Do Was Stay (Taylor's Version)   
..                                             ...   
270                                     Tim McGraw   
271                                Picture To Burn   
272    Teardrops On My Guitar - Radio Single Remix   
278                              Should've Said No   
280                                       Our Song   

                         album_title song_release_date  track      acou  \
0    1989 (Taylor's Version)[Deluxe]        10/27/2023      1  0.009420   
1    1989 (Taylor's Version)[Deluxe]        10/27/2023      2  0.088500   
2    1989 (Taylor's Version)[Deluxe]        10/27/2023      3  0.000421   
3    1989 (Taylor's Version)[Deluxe]        10/27/2023      4  0.000537   
4    1989 (Taylor's Version)[Deluxe]        10/27/2023      5  0.000656   
..                               ...               ...    ...       ...   
270                            Debut        10/24/2006      1  0.575000   
271                            Debut        10/24/2006      2  0.173000   
272                            Debut        10/24/2006      3  0.288000   
278                            Debut        10/24/2006      9  0.010300   
280                            Debut        10/24/2006     11  0.111000   

     dance   ener    instru    live   loud  speech     temp    val  pop  \
0    0.757  0.610  0.000037  0.3670 -4.840  0.0327  116.998  0.685   80   
1    0.733  0.733  0.000000  0.1680 -5.376  0.0670   96.057  0.701   80   
2    0.511  0.822  0.019700  0.0899 -4.785  0.0397   94.868  0.305   81   
3    0.545  0.885  0.000056  0.3850 -5.968  0.0447   92.021  0.206   80   
4    0.588  0.721  0.000000  0.1310 -5.579  0.0317   96.997  0.520   79   
..     ...    ...       ...     ...    ...     ...      ...    ...  ...   
270  0.580  0.491  0.000000  0.1210 -6.462  0.0251   76.009  0.425   68   
271  0.658  0.877  0.000000  0.0962 -2.098  0.0323  105.586  0.821   72   
272  0.621  0.417  0.000000  0.1190 -6.941  0.0231   99.953  0.289   69   
278  0.476  0.777  0.000000  0.1960 -3.771  0.0289  167.964  0.472   69   
280  0.668  0.672  0.000000  0.3290 -4.931  0.0303   89.011  0.539   77   

        dur album_id   song_id  days  tot_streams last_stream_update  \
0    212600   1989TV  1989TV01    13     35521714          11/9/2023   
1    231833   1989TV  1989TV02    13     40958068          11/9/2023   
2    231000   1989TV  1989TV03    13     47687102          11/9/2023   
3    235800   1989TV  1989TV04    13     37545128          11/9/2023   
4    193289   1989TV  1989TV05    13     31008534          11/9/2023   
..      ...      ...       ...   ...          ...                ...   
270  232106    DEBUT   DEBUT01  6225    112816077          11/9/2023   
271  173066    DEBUT   DEBUT02  6225    163382463          11/9/2023   
272  203040    DEBUT   DEBUT03  6225    191317513          11/9/2023   
278  242200    DEBUT   DEBUT09  6225    103848381          11/9/2023   
280  201106    DEBUT   DEBUT11  6225    280561783          11/9/2023   

     streams_n  top_song  top_song_flag  top_s_cnt                tv  alb_num  \
0      2732439  top song           True         14  Taylor's Version       14   
1      3150620  top song           True         14  Taylor's Version       14   
2      3668238  top song           True         14  Taylor's Version       14   
3      2888086  top song           True         14  Taylor's Version       14   
4      2385271  top song           True         14  Taylor's Version       14   
..         ...       ...            ...        ...               ...      ...   
270      18123  top song           True          1          original        1   
271      26246  top song           True          1          original        

In [33]:
# Update top song boolean value
# Check boolean counts


df['top_song_flag'].value_counts(dropna=False)

top_song_flag
True     192
False     93
Name: count, dtype: int64

In [34]:
# Update condition after top song value updates

condition = (df['top_song'] == 'top song')

In [35]:
# Update the boolean column ('column_name') to True where the condition is met
df.loc[condition, 'top_song_flag'] = True

In [36]:
df['top_song_flag'].value_counts(dropna=False)

top_song_flag
True     216
False     69
Name: count, dtype: int64

## Update count of top songs for albums

In [37]:
# Review counts- should match Billboard, using group by
# using song dataframe, group by album id and compare to actual billboard data

grouped_top = top.groupby('album_id')['top_song'].count()

In [38]:
grouped_top

album_id
1989       10
1989TV     22
DEBUT       5
EVER       15
FEAR       16
FEARTV      8
FOLK       16
LOVE       18
MIDN       22
RED        11
REDTV      27
REP         6
SPEAK      18
SPEAKTV    22
Name: top_song, dtype: int64

All updates complete

In [53]:
df.columns

Index(['song_title', 'album_title', 'song_release_date', 'track', 'acou',
       'dance', 'ener', 'instru', 'live', 'loud', 'speech', 'temp', 'val',
       'pop', 'dur', 'album_id', 'song_id', 'days', 'tot_streams',
       'last_stream_update', 'streams_n', 'top_song', 'top_song_flag',
       'top_s_cnt', 'tv', 'alb_num', 'album_announce_mo', 'alb_ann_day',
       'alb_ann_yr', 'album_announce_date', 'album_release_mo', 'alb_r_day',
       'alb_r_yr', 'album_release_date', 'w1_sales', 'top_alb_count',
       'top_alb_normal', 'alb_track_cnt', 'tvflag', 'sales_group',
       'tot_streams_group', 'alb_genre'],
      dtype='object')

In [54]:
# Check value counts

df['top_s_cnt'].value_counts(dropna=False)

top_s_cnt
11    30
10    26
12    23
14    22
13    22
4     22
3     20
5     19
2     19
7     18
9     17
8     17
6     15
1     15
Name: count, dtype: int64

In [56]:
df.groupby('album_id')['top_song'].median()

album_id
1989       19
1989TV     22
DEBUT      15
EVER       17
FEAR       19
FEARTV     26
FOLK       17
LOVE       18
MIDN       23
RED        22
REDTV      30
REP        15
SPEAK      20
SPEAKTV    22
Name: top_song, dtype: int64

will use unique top song counts from billboard df

In [58]:
# Update total top songs for albums

# Create a for loop to update top songs

alb_top_s_count= []

for value in df['album_id']:
    if value == "1989":
        alb_top_s_count.append(10)
    elif value == "DEBUT":
        alb_top_s_count.append(5)
    elif value == "REP":
        alb_top_s_count.append(6)
    elif value == "FEAR":
        alb_top_s_count.append(16)
    elif value == "LOVE":
        alb_top_s_count.append(18)
    elif value == "SPEAK":
        alb_top_s_count.append(17)
    elif value == "RED":
        alb_top_s_count.append(11)
    elif value == "FOLK":
        alb_top_s_count.append(16)
    elif value == "EVER":
        alb_top_s_count.append(15)
    elif value == "FEARTV":
        alb_top_s_count.append(8)
    elif value == "REDTV":
        alb_top_s_count.append(26)
    elif value == "MIDN":
        alb_top_s_count.append(22)
    elif value == "SPEAKTV":
        alb_top_s_count.append(22)
    elif value == "1989TV":
        alb_top_s_count.append(21)
    else:
        alb_top_s_count.append("unknown")

In [59]:
# update with new values

df['top_s_cnt'] = alb_top_s_count

In [62]:
df['top_s_cnt'].value_counts(dropna=False)

top_s_cnt
22    45
16    36
26    30
8     26
21    22
11    22
17    20
10    19
18    18
15    17
6     15
5     15
Name: count, dtype: int64

In [65]:
check_tops = df[['album_id', 'top_s_cnt']]

In [66]:
check_tops.groupby('album_id')['top_s_cnt'].median()

album_id
1989       10.0
1989TV     21.0
DEBUT       5.0
EVER       15.0
FEAR       16.0
FEARTV      8.0
FOLK       16.0
LOVE       18.0
MIDN       22.0
RED        11.0
REDTV      26.0
REP         6.0
SPEAK      17.0
SPEAKTV    22.0
Name: top_s_cnt, dtype: float64

Updated fully. Will export as updated df below

# 03. Update top song counts for album related dataframes
- album aggs (summary of all 14 albums)
- album scores (google search dataframe grouped by album-state) dataframes 

### Update albums df

In [51]:
df_alb_scores.head()

alb_state_key  alb_num album_id       state  state_num  \
0     1989, Alabama        5     1989     Alabama          1   
1      1989, Alaska        5     1989      Alaska          2   
2     1989, Arizona        5     1989     Arizona          3   
3    1989, Arkansas        5     1989    Arkansas          4   
4  1989, California        5     1989  California          5   

   st_group_alb_search_avg  alb_group_st_search_avg  alb_r_date record_type  \
0                     67.0                     67.0  2014-10-27    original   
1                     52.0                     67.0  2014-10-27    original   
2                     62.0                     67.0  2014-10-27    original   
3                     72.0                     67.0  2014-10-27    original   
4                     63.0                     67.0  2014-10-27    original   

   tvflag          st_alb_pop_group album_event  alb_track_cnt  wk1_sales  \
0   False    High search popularity        1989             16  1297000.0   
1   False  Medium search popularity        1989             16  1297000.0   
2   False  Medium search popularity        1989             16  1297000.0   
3   False    High search popularity        1989             16  1297000.0   
4   False  Medium search popularity        1989             16  1297000.0   

   tot_alb_plays  top_s_cnt  alb_streams_norm  top_alb_count  top_alb_n  \
0    220933979.5          5           66949.0            464          0   
1    220933979.5          5           66949.0            464          0   
2    220933979.5          5           66949.0            464          0   
3    220933979.5          5           66949.0            464          0   
4    220933979.5          5           66949.0            464          0   

   days_sr  alb_ann_day  alb_ann_yr  alb_r_day  alb_r_yr  acou_m  dance_m  \
0     3300           19        2014         15      2014    0.06     0.62   
1     3300           19        2014         15      2014    0.06     0.62   
2     3300           19        2014         15      2014    0.06     0.62   
3     3300           19        2014         15      2014    0.06     0.62   
4     3300           19        2014         15      2014    0.06     0.62   

   ener_m  instru_m  live_m  loud_m  speech_m  temp_m  val_m     dur_m  \
0    0.71       0.0     0.1   -5.92      0.05  120.98   0.49  230733.0   
1    0.71       0.0     0.1   -5.92      0.05  120.98   0.49  230733.0   
2    0.71       0.0     0.1   -5.92      0.05  120.98   0.49  230733.0   
3    0.71       0.0     0.1   -5.92      0.05  120.98   0.49  230733.0   
4    0.71       0.0     0.1   -5.92      0.05  120.98   0.49  230733.0   

     color alb_genre  
0  skyblue       pop  
1  skyblue       pop  
2  skyblue       pop  
3  skyblue       pop  
4  skyblue       pop

In [68]:
df_alb_scores.columns

Index(['alb_state_key', 'alb_num', 'album_id', 'state', 'state_num',
       'st_group_alb_search_avg', 'alb_group_st_search_avg', 'alb_r_date',
       'record_type', 'tvflag', 'st_alb_pop_group', 'album_event',
       'alb_track_cnt', 'wk1_sales', 'tot_alb_plays', 'top_s_cnt',
       'alb_streams_norm', 'top_alb_count', 'top_alb_n', 'days_sr',
       'alb_ann_day', 'alb_ann_yr', 'alb_r_day', 'alb_r_yr', 'acou_m',
       'dance_m', 'ener_m', 'instru_m', 'live_m', 'loud_m', 'speech_m',
       'temp_m', 'val_m', 'dur_m', 'color', 'alb_genre'],
      dtype='object')

In [69]:
check_alb_tops = df_alb_scores[['album_id', 'top_s_cnt']]

In [70]:
check_alb_tops.groupby('album_id')['top_s_cnt'].median()

album_id
1989        5.0
1989TV     14.0
DEBUT       1.0
EVER        9.0
FEAR        2.0
FEARTV     10.0
FOLK        8.0
LOVE        7.0
MIDN       12.0
RED         4.0
REDTV      11.0
REP         6.0
SPEAK       3.0
SPEAKTV    13.0
Name: top_s_cnt, dtype: float64

In [71]:
# Update with top values

# Create a for loop to update top songs

alb_top_s_count= []

for value in df_alb_scores['album_id']:
    if value == "1989":
        alb_top_s_count.append(10)
    elif value == "DEBUT":
        alb_top_s_count.append(5)
    elif value == "REP":
        alb_top_s_count.append(6)
    elif value == "FEAR":
        alb_top_s_count.append(16)
    elif value == "LOVE":
        alb_top_s_count.append(18)
    elif value == "SPEAK":
        alb_top_s_count.append(17)
    elif value == "RED":
        alb_top_s_count.append(11)
    elif value == "FOLK":
        alb_top_s_count.append(16)
    elif value == "EVER":
        alb_top_s_count.append(15)
    elif value == "FEARTV":
        alb_top_s_count.append(8)
    elif value == "REDTV":
        alb_top_s_count.append(26)
    elif value == "MIDN":
        alb_top_s_count.append(22)
    elif value == "SPEAKTV":
        alb_top_s_count.append(22)
    elif value == "1989TV":
        alb_top_s_count.append(21)
    else:
        alb_top_s_count.append("unknown")

In [72]:
df_alb_scores['top_s_cnt'] = alb_top_s_count

In [73]:
check_alb_tops = df_alb_scores[['album_id', 'top_s_cnt']]

In [74]:
check_alb_tops.groupby('album_id')['top_s_cnt'].median()

album_id
1989       10.0
1989TV     21.0
DEBUT       5.0
EVER       15.0
FEAR       16.0
FEARTV      8.0
FOLK       16.0
LOVE       18.0
MIDN       22.0
RED        11.0
REDTV      26.0
REP         6.0
SPEAK      17.0
SPEAKTV    22.0
Name: top_s_cnt, dtype: float64

In [75]:
df_alb_aggs

album_id                       album_title                tv  alb_num  \
0      1989             1989 (Deluxe Edition)          original      5.0   
1    1989TV   1989 (Taylor's Version)[Deluxe]  Taylor's Version     14.0   
2     DEBUT                             Debut          original      1.0   
3      EVER         evermore (deluxe version)          original      9.0   
4      FEAR         Fearless Platinum Edition          original      2.0   
5    FEARTV       Fearless (Taylor's Version)  Taylor's Version     10.0   
6      FOLK         folklore (deluxe version)          original      8.0   
7      LOVE                             Lover          original      7.0   
8      MIDN  Midnights (The Til Dawn Edition)          original     12.0   
9       RED              Red (Deluxe Edition)          original      4.0   
10    REDTV            Red (Taylor's Version)  Taylor's Version     11.0   
11      REP                        reputation          original      6.0   
12    SPEAK        Speak Now (Deluxe Edition)          original      3.0   
13  SPEAKTV      Speak Now (Taylor's Version)  Taylor's Version     13.0   

    alb_track_cnt  w1_trad_album_sales  tot_alb_plays  alb_streams_norm  \
0              16            1297000.0   2.209340e+08           66949.0   
1              22            1359000.0   1.462294e+09       112484136.0   
2              15              39000.0   1.164904e+09          187127.0   
3              17             154500.0   3.801517e+09         3584268.0   
4              19             592300.0   2.590572e+09          473068.0   
5              26             179000.0   3.776863e+09         4000902.0   
6              17             615000.0   6.156253e+09         5119556.0   
7              18             679000.0   8.029477e+09         5217324.0   
8              23            1140000.0   7.116052e+09        19032283.0   
9              22            1208000.0   3.204012e+09          794044.0   
10             30             369000.0   4.062936e+09         5588619.0   
11             15            1216000.0   6.546494e+09         2989259.0   
12             20            1047000.0   2.985802e+09          630440.0   
13             22             507000.0   1.524440e+09        12195506.0   

    avg_song_pop  top_s_cnt  top_alb_count  top_alb_normal      days_sr  \
0      67.157895        5.0          464.0        0.140606  3300.000000   
1      79.818182       14.0            1.0        0.076923    13.000000   
2      63.133333        1.0          284.0        0.045622  6225.000000   
3      72.764706        9.0          151.0        0.142486  1059.823529   
4      45.947368        2.0          261.0        0.047663  5476.000000   
5      71.615385       10.0          121.0        0.128178   944.000000   
6      73.058824        8.0          171.0        0.142322  1201.529412   
7      82.611111        7.0          219.0        0.142300  1539.000000   
8      69.608696       12.0           54.0        0.153507   366.869565   
9      47.863636        4.0          185.0        0.045849  4035.000000   
10     74.266667       11.0          103.0        0.141678   727.000000   
11     82.933333        6.0          267.0        0.121918  2190.000000   
12     52.800000        3.0          193.0        0.040752  4736.000000   
13     79.318182       13.0           17.0        0.136000   125.000000   

    alb_ann_day  alb_ann_yr  alb_r_day  alb_r_yr   acou_m  dance_m  ener_m  \
0          19.0      2014.0       15.0    2014.0  0.06000   0.6200  0.7100   
1           9.0      2023.0       27.0    2023.0  0.03860   0.6100  0.6785   
2          19.0      2006.0       24.0    2006.0  0.08680   0.5760  0.6720   
3          10.0      2020.0       11.0    2020.0  0.83000   0.5460  0.5170   
4          15.0      2008.0       11.0    2008.0  0.13000   0.5920  0.6360   
5          11.0      2021.0        9.0    2021.0  0.11100   0.5660  0.6250   
6          24.0      2020.0       24.0    2020.0  0.75000   0.59

In [76]:
check_alb_aggs_tops = df_alb_aggs[['album_id', 'top_s_cnt']]

In [77]:
check_alb_aggs_tops.groupby('album_id')['top_s_cnt'].median()

album_id
1989        5.0
1989TV     14.0
DEBUT       1.0
EVER        9.0
FEAR        2.0
FEARTV     10.0
FOLK        8.0
LOVE        7.0
MIDN       12.0
RED         4.0
REDTV      11.0
REP         6.0
SPEAK       3.0
SPEAKTV    13.0
Name: top_s_cnt, dtype: float64

In [79]:
# Update with top values

# Create a for loop to update top songs

alb_top_s_count= []

for value in df_alb_aggs['album_id']:
    if value == "1989":
        alb_top_s_count.append(10)
    elif value == "DEBUT":
        alb_top_s_count.append(5)
    elif value == "REP":
        alb_top_s_count.append(6)
    elif value == "FEAR":
        alb_top_s_count.append(16)
    elif value == "LOVE":
        alb_top_s_count.append(18)
    elif value == "SPEAK":
        alb_top_s_count.append(17)
    elif value == "RED":
        alb_top_s_count.append(11)
    elif value == "FOLK":
        alb_top_s_count.append(16)
    elif value == "EVER":
        alb_top_s_count.append(15)
    elif value == "FEARTV":
        alb_top_s_count.append(8)
    elif value == "REDTV":
        alb_top_s_count.append(26)
    elif value == "MIDN":
        alb_top_s_count.append(22)
    elif value == "SPEAKTV":
        alb_top_s_count.append(22)
    elif value == "1989TV":
        alb_top_s_count.append(21)
    else:
        alb_top_s_count.append("unknown")

In [80]:
df_alb_aggs['top_s_cnt'] = alb_top_s_count

In [81]:
check_alb_aggs_tops = df_alb_aggs[['album_id', 'top_s_cnt']]

In [82]:
check_alb_aggs_tops.groupby('album_id')['top_s_cnt'].median()

album_id
1989       10.0
1989TV     21.0
DEBUT       5.0
EVER       15.0
FEAR       16.0
FEARTV      8.0
FOLK       16.0
LOVE       18.0
MIDN       22.0
RED        11.0
REDTV      26.0
REP         6.0
SPEAK      17.0
SPEAKTV    22.0
Name: top_s_cnt, dtype: float64

# Check correlations after updates

In [86]:
# See if top song updates affected any correlations

numerical_alb_aggs = df_alb_aggs.select_dtypes(include='number')

In [87]:
numerical_alb_aggs.corr()

alb_num  alb_track_cnt  w1_trad_album_sales  \
alb_num              1.000000       0.554071             0.046707   
alb_track_cnt        0.554071       1.000000            -0.110127   
w1_trad_album_sales  0.046707      -0.110127             1.000000   
tot_alb_plays        0.164680       0.011324             0.040506   
alb_streams_norm     0.567630       0.187246             0.388202   
avg_song_pop         0.678014       0.014494            -0.056599   
top_s_cnt            0.626042       0.580633             0.076111   
top_alb_count       -0.779003      -0.627062             0.103164   
top_alb_normal       0.628949       0.143005            -0.112393   
days_sr             -0.957306      -0.446976            -0.046464   
alb_ann_day         -0.076402      -0.196743             0.144359   
alb_ann_yr           0.956037       0.450137             0.035107   
alb_r_day           -0.177810      -0.193394             0.353835   
alb_r_yr             0.956037       0.450137             0.035107   
acou_m               0.119272      -0.283731            -0.311140   
dance_m             -0.035013      -0.113410             0.494859   
ener_m              -0.257325      -0.014262             0.126449   
instru_m             0.166803      -0.123545            -0.052065   
live_m              -0.004415       0.022021            -0.179237   
loud_m              -0.352816       0.086311             0.007179   
speech_m             0.204674      -0.286177             0.525958   
temp_m               0.062139       0.110597            -0.115412   
val_m               -0.248571      -0.018884            -0.136538   
dur_m                0.053689       0.242810            -0.151603   

                     tot_alb_plays  alb_streams_norm  avg_song_pop  top_s_cnt  \
alb_num                   0.164680          0.567630      0.678014   0.626042   
alb_track_cnt             0.011324          0.187246      0.014494   0.580633   
w1_trad_album_sales       0.040506          0.388202     -0.056599   0.076111   
tot_alb_plays             1.000000         -0.196062      0.355081   0.151848   
alb_streams_norm         -0.196062          1.000000      0.330405   0.358370   
avg_song_pop              0.355081          0.330405      1.000000   0.193222   
top_s_cnt                 0.151848          0.358370      0.193222   1.000000   
top_alb_count            -0.176080         -0.527482     -0.309333  -0.645169   
top_alb_normal            0.461251         -0.091000      0.735404   0.289072   
days_sr                  -0.358792         -0.427916     -0.768557  -0.546743   
alb_ann_day               0.459819         -0.194920      0.131588  -0.175871   
alb_ann_yr                0.357583          0.420678      0.769116   0.536719   
alb_r_day                 0.093842          0.375787     -0.162896   0.061236   
alb_r_yr                  0.357583          0.420678      0.769116   0.536719   
acou_m                    0.337676         -0.146915      0.107021   0.060725   
dance_m                   0.609981          0.073878      0.231076   0.012587   
ener_m                   -0.697014          0.155193     -0.165035  -0.226497   
instru_m                  0.301566          0.001555      0.006773   0.096287   
live_m                   -0.117733         -0.072696      0.034269  -0.083720   
loud_m                   -0.633471         -0.034332     -0.323328  -0.162181   
speech_m                  0.584324          0.116756      0.522216  -0.085818   
temp_m                   -0.314418         -0.116318      0.045030  -0.039645   
val_m                    -0.320171         -0.144632     -0.027498  -0.265120   
dur_m                    -0.469326         -0.159243     -0.256273   0.160447   

                     top_alb_count  top_alb_normal   days_sr  alb_ann_day  \
alb_num                  -0.779003        0.628949 -0.957306    -0.076402   
alb_track_cnt            -0.627062        0.143005 -0.446976    -0.196743   
w1_trad_album_sales       0.103164     

In [88]:
# See if top song updates affected any correlations in the song information df 

numerical_songs = df.select_dtypes(include='number')

In [89]:
numerical_songs.corr()

track      acou     dance      ener    instru      live  \
track           1.000000  0.162762 -0.037213 -0.103154 -0.040061  0.003717   
acou            0.162762  1.000000 -0.115169 -0.697194  0.157630 -0.070554   
dance          -0.037213 -0.115169  1.000000  0.030465 -0.069636 -0.003571   
ener           -0.103154 -0.697194  0.030465  1.000000 -0.071019  0.088522   
instru         -0.040061  0.157630 -0.069636 -0.071019  1.000000 -0.040378   
live            0.003717 -0.070554 -0.003571  0.088522 -0.040378  1.000000   
loud           -0.181543 -0.709854 -0.044975  0.781243 -0.179001  0.083696   
speech          0.036444  0.123597  0.187526 -0.153323 -0.022708  0.263374   
temp            0.055894 -0.094232 -0.271033  0.148677 -0.004415 -0.012126   
val            -0.063633 -0.175339  0.322925  0.467342 -0.055276 -0.060420   
pop            -0.079275 -0.014763  0.038456  0.053140  0.001122 -0.103059   
dur             0.107374  0.028629 -0.298578 -0.111259 -0.030318 -0.126142   
days           -0.149398 -0.125539 -0.011085  0.112890 -0.067084  0.084410   
tot_streams    -0.343689 -0.029559  0.096125  0.034232 -0.002821 -0.073593   
streams_n      -0.024242 -0.078871  0.017191  0.037669 -0.007492 -0.033771   
top_s_cnt       0.173441  0.027549 -0.013263 -0.078718  0.037189 -0.085417   
alb_num         0.175419  0.033988 -0.011175 -0.048338  0.059882 -0.055187   
alb_ann_day    -0.029585  0.138237  0.131404 -0.247246  0.054186  0.018408   
alb_ann_yr      0.148719  0.122879  0.007232 -0.109767  0.063741 -0.085282   
alb_r_day      -0.091221  0.039577  0.111263 -0.051086 -0.004600  0.009553   
alb_r_yr        0.148719  0.122879  0.007232 -0.109767  0.063741 -0.085282   
w1_sales       -0.051742 -0.166198  0.288104  0.010484 -0.011160  0.110343   
top_alb_count  -0.164930  0.007816  0.088747  0.018555 -0.050987  0.103557   
top_alb_normal  0.114069  0.255446  0.017810 -0.191473  0.065638 -0.010514   
alb_track_cnt   0.324203 -0.215184 -0.036442  0.068838 -0.028942 -0.016280   

                    loud    speech      temp       val       pop       dur  \
track          -0.181543  0.036444  0.055894 -0.063633 -0.079275  0.107374   
acou           -0.709854  0.123597 -0.094232 -0.175339 -0.014763  0.028629   
dance          -0.044975  0.187526 -0.271033  0.322925  0.038456 -0.298578   
ener            0.781243 -0.153323  0.148677  0.467342  0.053140 -0.111259   
instru         -0.179001 -0.022708 -0.004415 -0.055276  0.001122 -0.030318   
live            0.083696  0.263374 -0.012126 -0.060420 -0.103059 -0.126142   
loud            1.000000 -0.355841  0.111320  0.304661 -0.046836  0.093525   
speech         -0.355841  1.000000 -0.001050  0.087632  0.018712 -0.349132   
temp            0.111320 -0.001050  1.000000  0.046200  0.014652  0.010854   
val             0.304661  0.087632  0.046200  1.000000 -0.007948 -0.323466   
pop            -0.046836  0.018712  0.014652 -0.007948  1.000000 -0.065471   
dur             0.093525 -0.349132  0.010854 -0.323466 -0.065471  1.000000   
days            0.263770 -0.040494  0.002780  0.062919 -0.734954 -0.000805   
tot_streams    -0.014524  0.054268 -0.098651  0.075659  0.277045 -0.152814   
streams_n      -0.028023 -0.004662  0.015827 -0.017413  0.161880 -0.060617   
top_s_cnt      -0.075984 -0.062725 -0.014393 -0.032351  0.228709  0.156095   
alb_num        -0.167885 -0.002744  0.009554 -0.065564  0.658188  0.036908   
alb_ann_day    -0.409935  0.167521 -0.093417 -0.167525  0.095966 -0.305746   
alb_ann_yr     -0.256514  0.037413 -0.000497 -0.060712  0.734868  0.003440   
alb_r_day      -0.130741  0.011210 -0.056183  0.021364 -0.159874 -0.224221   
alb_r_yr       -0.256514  0.037413 -0.000497 -0.060712  0.734868  0.003440   
w1_sales       -0.143671  0.211833 -0.040151 -0.060035 -0.107003 -0.154139   
top_alb_count   0.007274  0.233394  0.002615  0.068719 -0.323822 -0.158495   
top_alb_normal -0.331459  0.173789 -0.020351 -0.073811  0.673879 -0.078653   
alb_track_cnt   0.077434 -0.107590  0.01

Track is the strongest relationship in sound characteristics with top song count- 0.173, no other significant relationships

In [90]:
df.columns

Index(['song_title', 'album_title', 'song_release_date', 'track', 'acou',
       'dance', 'ener', 'instru', 'live', 'loud', 'speech', 'temp', 'val',
       'pop', 'dur', 'album_id', 'song_id', 'days', 'tot_streams',
       'last_stream_update', 'streams_n', 'top_song', 'top_song_flag',
       'top_s_cnt', 'tv', 'alb_num', 'album_announce_mo', 'alb_ann_day',
       'alb_ann_yr', 'album_announce_date', 'album_release_mo', 'alb_r_day',
       'alb_r_yr', 'album_release_date', 'w1_sales', 'top_alb_count',
       'top_alb_normal', 'alb_track_cnt', 'tvflag', 'sales_group',
       'tot_streams_group', 'alb_genre'],
      dtype='object')

In [97]:
# Check correlation using the boolean flag

df_top_flag = df[['track', 'acou',
                   'dance', 'ener', 
                  'instru', 'live', 
                  'loud', 'speech', 
                  'temp', 'val',
                   'dur', 'days', 
                  'tot_streams', 'streams_n', 
                  'top_song_flag',
                  'top_s_cnt', 'alb_num']]

In [98]:
df_top_flag.corr()

track      acou     dance      ener    instru      live  \
track          1.000000  0.162762 -0.037213 -0.103154 -0.040061  0.003717   
acou           0.162762  1.000000 -0.115169 -0.697194  0.157630 -0.070554   
dance         -0.037213 -0.115169  1.000000  0.030465 -0.069636 -0.003571   
ener          -0.103154 -0.697194  0.030465  1.000000 -0.071019  0.088522   
instru        -0.040061  0.157630 -0.069636 -0.071019  1.000000 -0.040378   
live           0.003717 -0.070554 -0.003571  0.088522 -0.040378  1.000000   
loud          -0.181543 -0.709854 -0.044975  0.781243 -0.179001  0.083696   
speech         0.036444  0.123597  0.187526 -0.153323 -0.022708  0.263374   
temp           0.055894 -0.094232 -0.271033  0.148677 -0.004415 -0.012126   
val           -0.063633 -0.175339  0.322925  0.467342 -0.055276 -0.060420   
dur            0.107374  0.028629 -0.298578 -0.111259 -0.030318 -0.126142   
days          -0.149398 -0.125539 -0.011085  0.112890 -0.067084  0.084410   
tot_streams   -0.343689 -0.029559  0.096125  0.034232 -0.002821 -0.073593   
streams_n     -0.024242 -0.078871  0.017191  0.037669 -0.007492 -0.033771   
top_song_flag -0.212150 -0.010557 -0.005184  0.066288  0.048740 -0.068292   
top_s_cnt      0.173441  0.027549 -0.013263 -0.078718  0.037189 -0.085417   
alb_num        0.175419  0.033988 -0.011175 -0.048338  0.059882 -0.055187   

                   loud    speech      temp       val       dur      days  \
track         -0.181543  0.036444  0.055894 -0.063633  0.107374 -0.149398   
acou          -0.709854  0.123597 -0.094232 -0.175339  0.028629 -0.125539   
dance         -0.044975  0.187526 -0.271033  0.322925 -0.298578 -0.011085   
ener           0.781243 -0.153323  0.148677  0.467342 -0.111259  0.112890   
instru        -0.179001 -0.022708 -0.004415 -0.055276 -0.030318 -0.067084   
live           0.083696  0.263374 -0.012126 -0.060420 -0.126142  0.084410   
loud           1.000000 -0.355841  0.111320  0.304661  0.093525  0.263770   
speech        -0.355841  1.000000 -0.001050  0.087632 -0.349132 -0.040494   
temp           0.111320 -0.001050  1.000000  0.046200  0.010854  0.002780   
val            0.304661  0.087632  0.046200  1.000000 -0.323466  0.062919   
dur            0.093525 -0.349132  0.010854 -0.323466  1.000000 -0.000805   
days           0.263770 -0.040494  0.002780  0.062919 -0.000805  1.000000   
tot_streams   -0.014524  0.054268 -0.098651  0.075659 -0.152814 -0.040420   
streams_n     -0.028023 -0.004662  0.015827 -0.017413 -0.060617 -0.204315   
top_song_flag  0.056607 -0.096033 -0.056026  0.047875  0.090653 -0.246723   
top_s_cnt     -0.075984 -0.062725 -0.014393 -0.032351  0.156095 -0.532730   
alb_num       -0.167885 -0.002744  0.009554 -0.065564  0.036908 -0.960768   

               tot_streams  streams_n  top_song_flag  top_s_cnt   alb_num  
track            -0.343689  -0.024242      -0.212150   0.173441  0.175419  
acou             -0.029559  -0.078871      -0.010557   0.027549  0.033988  
dance             0.096125   0.017191      -0.005184  -0.013263 -0.011175  
ener              0.034232   0.037669       0.066288  -0.078718 -0.048338  
instru           -0.002821  -0.007492       0.048740   0.037189  0.059882  
live             -0.073593  -0.033771      -0.068292  -0.085417 -0.055187  
loud             -0.014524  -0.028023       0.056607  -0.075984 -0.167885  
speech            0.054268  -0.004662      -0.096033  -0.062725 -0.002744  
temp             -0.098651   0.015827      -0.056026  -0.014393  0.009554  
val               0.075659  -0.017413       0.047875  -0.032351 -0.065564  
dur              -0.152814  -0.060617       0.090653   0.156095  0.036908  
days             -0.040420  -0.204315      -0.246723  -0.532730 -0.960768  
tot_streams       1.000000   0.085093       0.250421  -0.103760 -0.076175  
streams_n         0.085093   1.000000       0.114745   0.153831  0.266742  
top_song_flag     0.250421   0.114745       1.000000   0.529228  0.267336  
top_s_cnt        -0.103760  

Total streams has the strongest relationship with top song indicator (0.25), days since release (-0.24), and track (-.212150)

# 04. Data exports

In [67]:
# Export song info dataframe with multiple column updates


df.to_csv(os.path.join(path, '02 Manipulated Data', 'song_stream_alb_final_2.csv'))


In [78]:
# Export album scores dataframe with only top song updates

df_alb_scores.to_csv(os.path.join(path, '02 Manipulated Data', 'final_album_scores_3.csv'))

In [83]:
# Export album aggs dataframe with only top song updates

df_alb_aggs.to_csv(os.path.join(path, '02 Manipulated Data', 'alb_aggs_final_2.csv'))